### PR TITLE
[Graph memory plan] Support nested tuples

### PR DESCRIPTION
**Pull request by masahi**
The fix for the issue https://discuss.tvm.apache.org/t/graph-plan-memory-doesnt-support-nested-tuples/8278

A token is created for each tensor in a nested tuple. Graph runtime doesn't deal with tuples, so nested tuples are all flattened (`runtime.get_num_outputs()` and `runtime.get_output(i)` don't take tuples or nesting into account).  

please review @tqchen @zhiics  (cc @mbaret @manupa-arm )
I didn't look into memory planning in detail, but I hope this is correct.

### CODE REVIEW

File: src/relay/backend/graph_plan_memory.cc [hunk 1]
```diff
@@ -82,9 +82,9 @@ class StorageAllocaBaseVisitor : public ExprVisitor {
   void VisitExpr_(const TupleNode* op) final {
     std::vector<StorageToken*> fields;
     for (Expr field : op->fields) {
-      auto tok = GetToken(field);
-      ICHECK_EQ(tok.size(), 1U);
-      fields.push_back(tok[0]);
+      VisitExpr(field);
```
Comment (zhiics) [inline src/relay/backend/graph_plan_memory.cc:85]: Isn't field visited in GetToken?

Comment (masahi) [reply src/relay/backend/graph_plan_memory.cc]: thanks, you are right. fixed

**Review by zhiics — APPROVED** [review]

File: src/relay/backend/graph_plan_memory.cc [hunk 2]
```diff
@@ -82,9 +82,8 @@ class StorageAllocaBaseVisitor : public ExprVisitor {
   void VisitExpr_(const TupleNode* op) final {
     std::vector<StorageToken*> fields;
     for (Expr field : op->fields) {
-      auto tok = GetToken(field);
-      ICHECK_EQ(tok.size(), 1U);
-      fields.push_back(tok[0]);
+      auto tokens = GetToken(field);
```
Comment (jroesch) [inline src/relay/backend/graph_plan_memory.cc:85]: Does this handle 2 levels of nesting?

File: tests/python/relay/test_backend_graph_runtime.py [hunk 3]
```diff
@@ -184,10 +184,35 @@ def unit_numpy(X, W):
             tvm.testing.assert_allclose(out, ref, rtol=1e-5, atol=1e-5)
 
 
+def test_compile_nested_tuples():
+    x = relay.var("x", shape=(10,))
+    x1 = x + relay.const(1.0)
+    x2 = x1 + relay.const(1.0)
+    x3 = x2 + relay.const(1.0)
+    out = relay.Tuple([x1, relay.Tuple([x2, x3])])
```
Comment (jroesch) [inline tests/python/relay/test_backend_graph_runtime.py:192]: For example what if x2 is also a tuple.

File: src/relay/backend/graph_plan_memory.cc (hunk 2 above)
Comment (masahi) [reply src/relay/backend/graph_plan_memory.cc:85]: yes, there is `VisitExpr(field)` inside `GetToken`, where nesting is recursively visited and token generated for each tensor in the nest.

File: tests/python/relay/test_backend_graph_runtime.py (hunk 3 above)
Comment (masahi) [reply tests/python/relay/test_backend_graph_runtime.py]: just updated this test to verify two level nesting works.

**Review by mbaret — APPROVED** [review]
This change lgtm, but I have a question relating to the original discuss post. I've been looking at the compile_engine recently and noticed a number of TODOs there with the comment 'Allow recursive tuple', see for instance: https://github.com/apache/incubator-tvm/blob/3222cad9464cadafcab29dbfb5cf149bf083913f/src/relay/backend/compile_engine.cc#L118

It also seems to me that there's logic in `graph_runtime_codegen.cc` that flattens tuples assuming they're not recursive. Therefore, I'm a bit surprised it was only graph_plan_memory that was blocking you.

**Comment by masahi** [comment]
@mbaret yes, so this patch suffices for my simple use case of returning nested tuples from the `main` function. But for use cases like BYOC where the `checked_type` of `CallNode->op` can be a nested tuple, it seems I need further fixes. 

For example, I've just tried compiling this graph containing a sub function with nested tuple output, and it doesn't compile. I'll work on the fix, that should enable removing the tuple flatting in the BYOC partitioning pass @manupa-arm  

```
#[version = "0.0.5"]
def @func0(%x: Tensor[(10, 10), uint8]) -> (Tensor[(5, 10), uint8], (Tensor[(5, 10), uint8], Tensor[(5, 10), uint8])) {
  %0 = split(%x, indices_or_sections=2) /* ty=(Tensor[(5, 10), uint8], Tensor[(5, 10), uint8]) */;
  %1 = %0.0;
  %2 = %0.1;
  %3 = %0.0;
  %4 = abs(%3) /* ty=Tensor[(5, 10), uint8] */;
  %5 = (%2, %4);
  (%1, %5)
}

def @main(%a: Tensor[(10, 10), uint8]) -> (Tensor[(5, 10), uint8], (Tensor[(5, 10), uint8], Tensor[(5, 10), uint8])) {
  @func0(%a) /* ty=(Tensor[(5, 10), uint8], (Tensor[(5, 10), uint8], Tensor[(5, 10), uint8])) */
}
```

**Comment by masahi** [comment]
@tqchen @zhiics @jroesch @mbaret Supporting nested tuples in `CallNode->op` return type required making non trivial changes to multiple files in the commit https://github.com/apache/incubator-tvm/pull/6809/commits/d384062f4e2c38755cb9ae190d218d366cec906f

Maybe it would help BYOC use cases, but otherwise I don't know if we want to support this in practice. So I'm happy to revert that change and only support returning nested tuples from `main` function.

**Comment by zhiics** [comment]
I am actually okay without the new change

**Comment by mbaret** [comment]
I'm fine to leave the further changes out of this PR (especially as it's titled with graph plan memory). If it looks more complicated, that may need to be the subject of an RFC.

**Comment by masahi** [comment]
ok reverted

**Comment by masahi** [comment]
Thanks @zhiics @jroesch @mbaret

**Pull request merged**